micro-rdk/src/common/log.rs

```diff
@@ -2,40 +2,189 @@ use crate::{
+
+// We've chosen a size of 150 for the buffer due to a roughly observed maximum of 200 bytes per log message and
+// a desire to restrict the total amount of space for the cache to 30KB without losing logs to overwriting
+// overwriting the ring buffer between uploads. The consequence is that, when the device is offline, we will
+// cache the last 150 logs.
+pub(crate) fn get_log_buffer() -> &'static AsyncMutex<LogBufferType> {
+    static LOG_BUFFER: OnceLock<AsyncMutex<LogBufferType>> = OnceLock::new();
+    LOG_BUFFER.get_or_init(|| AsyncMutex::new(LocalRb::new(150)))
+}
+
+pub(crate) struct LogUploadTask {}
+
+impl PeriodicAppClientTask for LogUploadTask {
+    fn get_default_period(&self) -> std::time::Duration {
+        Duration::from_secs(1)
+    }
+    fn name(&self) -> &str {
+        "LogUpload"
+    }
+    fn invoke<'b, 'a: 'b>(
+        &'a mut self,
+        app_client: &'b AppClient,
+    ) -> std::pin::Pin<
+        Box<dyn std::future::Future<Output = Result<Option<Duration>, AppClientError>> + 'b>,
+    > {
+        Box::pin(async move {
+            let entries: Vec<LogEntry> = {
+                let mut logs = get_log_buffer().lock().await;
+                logs.pop_iter()
+                    .map(|log_entry| log_entry.get_time_corrected_entry())
+                    .collect()
+            };
+            if entries.is_empty() {
+                Ok(None)
+            } else {
+                app_client.push_logs(entries).await.map(|_| None)
+            }
+        })
+    }
+}
+
+pub trait ViamLogAdapter {
+    fn before_log_setup(&self);
+    fn get_level_filter(&self) -> ::log::LevelFilter;
+    fn new() -> Self;
+}
+
+// ViamLogger is a wrapper around an existing logger that stores a copy into LOG_BUFFER for later
+// upload to the cloud. The existing logger should satisfy log::Log and the ViamLogAdapter
+// trait and then by initialized using this function at the start of main
+pub fn initialize_logger<T: ::log::Log + ViamLogAdapter + 'static>() {
+    let inner = T::new();
+    let logger = ViamLogger::new(inner);
+    let filter = logger.level_filter();
+    logger.before_log_setup();
+    let _ = ::log::set_boxed_logger(Box::new(logger));
+    ::log::set_max_level(filter)
+}
+
+struct ViamLogger<L>(L);
+
+impl<L> ViamLogger<L>
+where
+    L: ::log::Log + ViamLogAdapter,
+{
+    fn new(inner: L) -> Self {
+        Self(inner)
+    }
+
+    fn before_log_setup(&self) {
+        self.0.before_log_setup()
+    }
+
+    fn level_filter(&self) -> ::log::LevelFilter {
+        self.0.get_level_filter()
+    }
+}
+
+impl<L> ::log::Log for ViamLogger<L>
+where
+    L: ::log::Log + ViamLogAdapter,
+{
+    fn enabled(&self, metadata: &log::Metadata) -> bool {
+        self.0.enabled(metadata)
+    }
+
+    fn flush(&self) {
+        self.0.flush()
+    }
+
+    fn log(&self, record: &log::Record) {
+        if self.enabled(record.metadata()) {
+            self.0.log(record);
```

Comment:
same here this would format the string twice.

Reply:
I'm not sure how to avoid this aside from completely recreating the logic in `EspLogger`. The creation of the formatted string is down entirely in `EspLogger::log` without delegation to another function. Additionally, the formatting of the string is completely different from what app.viam.com is expecting of us, so we would like have to format the string twice anyway